### ISSUE TITLE
Add code coverage for unit tests?

### ISSUE THREAD

**Issue by mkschreder** [warmcat/libwebsockets#617]
How about adding coveralls.io badge?

**Comment by JoakimSoderberg** [comment]
There are no unit tests so I haven't added that. But maybe running a test server + client would be an idea, to see how much of the code is used, as a test.

I've made a project to add coveralls to C/C++ projects using cmake:
https://github.com/JoakimSoderberg/coveralls-cmake

If someone feels it worthwhile adding that.

I prefer codecov though, a lot less fiddling and configuring (no need to write your own cmake project), and much more stable from my experience:
https://codecov.io/

**Comment by lws-team** [comment]
Lws has test apps that interoperate with each other, browsers and servers, a script that checks for handling some common attack vectors and Autobahn.

If the OP is asking if we'd like to have tests scripted and run in ci, sure, although running autobahn every time the repo changes will probably warm the planet a bit.  And I guess once it's there if it's not hard to use the number of tests would increase (although no cloud ci is going to allow coverage of, eg, the ESP8266 support...) But it's not at the top of my list to work on personally.

The missing bit here is more test scripts I think, not so much a cloud service.  Thinking about the new code like lwsgs, I test it in a browser because he has forms and such.

**Comment by mkschreder** [comment]
How do you actually make sure that all of your code is being tested if you
don't run coverage?

**Comment by lws-team** [comment]
Hm this is a "have you stopped beating your wife" kind of question isn't it.

Are you offering to help with this?

**Comment by JoakimSoderberg** [comment]
@mkschreder I'm also a fan of unit tests, but that does not mean it is the only way to make sure code is tested and works. Look at what one could call the most popular piece of software around, the Linux kernel. No unit tests there :)

The thing that is closest to a unit test in libwebsockets I guess would be the fact that the (correct me if I am wrong @lws-team) echo test client/server passes the autobahn test suite. One idea would be to run those, both through valgrind + calculate coverage.

I played around with this in my own hobby websocket client project:
https://github.com/JoakimSoderberg/libws/blob/master/.travis-ci-autobahn.sh

**Comment by mkschreder** [comment]
@Joakim, I agree that it is not the only way, but my personal experience is
that it forces me to rethink the code and actually design the whole thing
better. Because in order to test the websocket protocol I would probably
have to split the protocol itself from the transfer (ie the socket) and
make sure that the protocol stack works with plain char buffers before
running it on the network. One benefit of coverage is that it will pinpoint
the code paths that do not run during the tests. For example, if you have
an invalid memory access that will only occure under a specific condition -
this will probably not occur during a test unless you also do code coverage
and actually see that it is not being tested. Seeing that on a summary page
with all the files and numbers of exactly how much code is actually being
tested is such a valuable tool I don't really know of an alternative that
would do better (and no, only since one echo test works in autobahn does
not mean that all code is being tested - in fact, it could be that only 8%
of all code is being tested, or 40% or some other arbitrary number. You
don't know unless you actually run coverage).

Andy, do you want me to add basic travis coverage support to the
.travis.yml and makefiles? As far as designing unit tests, I think you are
probably the best person to do this since you know this library better than
anyone else.

**Comment by lws-team** [comment]
Yes that's right... in fact I have a checklist in the tree of things to confirm before each release and it includes autobahn, attack.sh and go through the test apps.

I think three separate things have gotten mixed up here

 - enabling more scriptable tests

 - assessing a coverage number for whatever tests are done, including manual ones

 - doing stuff in CI

Doing unrepresentative and partial testing in CI is not useful, when it delivers a falsely low number, so that's the least interesting thing.

Can I ask you from what direction your interest is in lws test coverage?  You're using lws yourself?  Something else?

lws passes Autobahn... that's a fact...  lws does a lot of things that Autobahn doesn't know or care about, so the percentage of coverage resulting from running Autobahn on lws is a function of Autobahn as much as lws.  I dunno what is the value of discovering what that number is.  As I mentioned nothing's going to autotest ESP8266 without an ESP8266... lws should get beaten over the head with that on test coverage numbers because it supports more platforms than anything similar?

Getting good coverage numbers is a major project because it requires testing error paths not just function.  Right now I don't have a reason to spend a lot of time on that.

**Comment by lws-team** [comment]
Ignoring the CI / scriptable thing, if nothing else this discussion made me curious what kind of coverage my list of "empirical stimulus" tests actually give.

I added support for gcc's gcov locally which is very cool, I'll look further at it tomorrow.  gcov solves most of the problems of coverage testing with the current structure, including ESP8266 type issues.

**Comment by mkschreder** [comment]
Very cool.

You were curious why I brought this up.. I use libwesockets for websocket
support in orangerpcd (which is official backend server for juci project
https://github.com/mkschreder/juci.git) It is kindof important for me that
it is stable because it fulfills such a vital function. I'm also
progressively adding better coverage and more unit tests to as much of
backend functionality as possible.

On 31 Aug 2016 15:25, "Andy Green" <notifications@github.com> wrote:

Ignoring the CI / scriptable thing, if nothing else this discussion made me
curious what kind of coverage my list of "empirical stimulus" tests
actually give.

I added support for gcc's gcov locally which is very cool, I'll look
further at it tomorrow. gcov solves most of the problems of coverage
testing with the current structure, including ESP8266 type issues.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/warmcat/libwebsockets/issues/617#issuecomment-243762694>,
or mute the thread

**Comment by lws-team** [comment]
Fair enough.... if it's any comfort lws (and lwsws) runs my internet-facing web services for most of a year now, so I am very much eating my own dogfood.  I update it to master when master has something new that stabilizes, the longest between updates was a month of uptime.  That's running all the features like CGI, plugins, mounts, generic sessions and so on.

It doesn't "prove" there are no problems (just no problems that got hit) but it does put a reasonably strict limit on how unstable it can be.

**Comment by mkschreder** [comment]
Haven't had any problems so far myself either.

**Comment by JoakimSoderberg** [comment]
@lws-team my project here runs gcov in the end https://github.com/JoakimSoderberg/coveralls-cmake

Here's an example project to set it up using cmake:
https://github.com/JoakimSoderberg/coveralls-cmake-example/

What is nice is that one can specify only the lib srcs to get coverage for the core parts (and not include all the test code itself).

Codecov also has a chrome plugin that overlays coverage data in diffs on github

**Comment by lws-team** [comment]
Right... that's what I meant by it 'solving most of the issues', it can aggregate coverage from the test apps and autobahn.  It also sets "100%" as "100% of the built code" which, at the cost of making the numbers very specific to the platform + config, eliminates getting burdened with nonplatform code being permanently out of coverage.

It seems to have some quirks though and googling the only way I found to get results from it is run it from a build dir four levels down... results are not updated from attack.sh somehow either I will look at that thismorning.

Thanks for the chrome plugin hint, I made a script that postprocesses and sorts the results, and does the math to calculate total covered lines vs total lines as reported by running gcov.

**Comment by JoakimSoderberg** [comment]
`gcov` does all the work, generates the coverage and so on and creates the actual report. You can also get an HTML version based on that. The issue with this as you say is that it only reports coverage for what you built that particular time. So you don't get the "overall picture". Also it includes coverage of the actual files running the unit tests / tests by default.

All of these reports are local to your computer, and no use here on github of course :)

However, there are 2 coverage reporting services online. One that @mkschreder named https://coveralls.io/ I wrote a CMake project for this to make everything easy to use and upload. In that project, I have made it so that you get to specify **which source files you care about getting coverage reports for**. `gcov` will still calculate it for all the files that get compiled, but when uploading the report only the ones you specified are included.

So once you have uploaded the coverage report, you get a report like this: https://coveralls.io/github/JoakimSoderberg/cargo

They have a nice HTML version of the files with the coverage info and so on. However, coveralls (unless they have changed this since I last checked) still have the problem you mentioned, they only allow 1 report per build... if you have multiple `#ifdef`s and want to recompile multiple times and then combine the coverage support, that is not supported.

So I prefer this other service https://codecov.io/ since first of all their upload procedure is awesome. You can pretty much run coverage in any language you want, run their script, and they will find and upload any relevant files (and on their website you can exclude certain files as I mentioned being able to do in my CMake thing for coveralls).

Some features codecov supports:
* For one build, combine multiple reports from multiple builds into one picture (decided by git hash)
* Browser plugin for chrome and firefox showing the overlay on github with coverage info in diffs and the source code
* Has that simple script (that if I knew existed never would've bothered with my cmake project for coveralls).

... so yea, end of maybe unecessary explination :)

**Comment by mkschreder** [comment]
Very good explanation @Joakim Söderberg.

**Comment by taeasy** [comment]
--2016-11-05 02:52:37--  https://codeload.github.com/warmcat/libwebsockets/tar.gz/v2.0.3?/libwebsockets-2.0.3.tar.gz
Resolving codeload.github.com (codeload.github.com)... 192.30.253.121, 192.30.253.120
Connecting to codeload.github.com (codeload.github.com)|192.30.253.121|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/x-gzip]
Saving to: 'STDOUT'

-                                                           [                                                                                  <=>                                            ]   2.86M   197KB/s   in 21s    

2016-11-05 02:52:59 (142 KB/s) - written to stdout [3000164]

. /home/main/x86/include/shell.sh; gzip -dc /home/main/x86/dl/libwebsockets-2.0.3.tar.gz | tar -C /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.. -xf - 
touch /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.prepared_03455b160277fb6bf2fdc8ebbe4e2378
rm -f /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/stamp/.libwebsockets.cyassl_installed
(cd /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3; CFLAGS="-Os -pipe -march=pentium4 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -iremap /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3:libwebsockets-2.0.3 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " CXXFLAGS="-Os -pipe -march=pentium4 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -iremap /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3:libwebsockets-2.0.3 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro " LDFLAGS="-L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/lib -znow -zrelro " cmake -DCMAKE_SYSTEM_NAME=Linux -DCMAKE_SYSTEM_VERSION=1 -DCMAKE_SYSTEM_PROCESSOR=i386 -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS_RELEASE="-DNDEBUG" -DCMAKE_CXX_FLAGS_RELEASE="-DNDEBUG" -DCMAKE_C_COMPILER="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc" -DCMAKE_C_COMPILER_ARG1="" -DCMAKE_CXX_COMPILER="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-g++" -DCMAKE_CXX_COMPILER_ARG1="" -DCMAKE_ASM_COMPILER="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc" -DCMAKE_ASM_COMPILER_ARG1="" -DCMAKE_EXE_LINKER_FLAGS:STRING="-L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/lib -znow -zrelro" -DCMAKE_MODULE_LINKER_FLAGS:STRING="-L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/lib -znow -zrelro -Wl,-Bsymbolic-functions" -DCMAKE_SHARED_LINKER_FLAGS:STRING="-L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/lib -znow -zrelro -Wl,-Bsymbolic-functions" -DCMAKE_AR="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc-ar" -DCMAKE_NM="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc-nm" -DCMAKE_RANLIB="/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc-ranlib" -DCMAKE_FIND_ROOT_PATH="/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr;/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15" -DCMAKE_FIND_ROOT_PATH_MODE_PROGRAM=BOTH -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY -DCMAKE_FIND_ROOT_PATH_MODE_INCLUDE=ONLY -DCMAKE_STRIP=: -DCMAKE_INSTALL_PREFIX=/usr -DDL_LIBRARY=/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15 -DCMAKE_PREFIX_PATH=/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15 -DCMAKE_SKIP_RPATH=TRUE -DLWS_IPV6=ON -DLWS_WITHOUT_TESTAPPS=ON -DLWS_OPENSSL_CLIENT_CERTS=/etc/ssl/certs -DLWS_OPENSSL_SUPPORT=ON -DLWS_WITH_SSL=ON -DLWS_USE_CYASSL=ON -DLWS_CYASSL_LIBRARIES=/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib/libcyassl.so -DLWS_CYASSL_INCLUDE_DIRS=/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include . )
-- The C compiler identification is GNU 5.3.0
-- Check for working C compiler: /home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc
-- Check for working C compiler: /home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/bin/i486-openwrt-linux-musl-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMAKE_TOOLCHAIN_FILE=''
-- Found Git: /home/main/x86/staging_dir/host/bin/git  
fatal: No annotated tags can describe 'cabb0c506526fe468e4aa985086e50c440192122'.
However, there were unannotated tags: try --tags.
Git commit hash: main@debian-
-- Performing Test LWS_HAVE_inline
-- Performing Test LWS_HAVE_inline - Success
-- Performing Test LWS_HAVE___inline__
-- Performing Test LWS_HAVE___inline__ - Success
-- Performing Test LWS_HAVE___inline
-- Performing Test LWS_HAVE___inline - Success
-- Looking for bzero
-- Looking for bzero - found
-- Looking for fork
-- Looking for fork - found
-- Looking for getenv
-- Looking for getenv - found
-- Looking for malloc
-- Looking for malloc - found
-- Looking for memset
-- Looking for memset - found
-- Looking for realloc
-- Looking for realloc - found
-- Looking for socket
-- Looking for socket - found
-- Looking for strerror
-- Looking for strerror - found
-- Looking for vfork
-- Looking for vfork - found
-- Looking for execvpe
-- Looking for execvpe - found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for snprintf
-- Looking for snprintf - found
-- Looking for _snprintf
-- Looking for _snprintf - not found
-- Looking for _vsnprintf
-- Looking for _vsnprintf - not found
-- Looking for getloadavg
-- Looking for getloadavg - found
-- Looking for dlfcn.h
-- Looking for dlfcn.h - found
-- Looking for fcntl.h
-- Looking for fcntl.h - found
-- Looking for in6addr.h
-- Looking for in6addr.h - not found
-- Looking for inttypes.h
-- Looking for inttypes.h - found
-- Looking for memory.h
-- Looking for memory.h - found
-- Looking for netinet/in.h
-- Looking for netinet/in.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stdlib.h
-- Looking for stdlib.h - found
-- Looking for strings.h
-- Looking for strings.h - found
-- Looking for string.h
-- Looking for string.h - found
-- Looking for sys/prctl.h
-- Looking for sys/prctl.h - found
-- Looking for sys/socket.h
-- Looking for sys/socket.h - found
-- Looking for sys/stat.h
-- Looking for sys/stat.h - found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for unistd.h
-- Looking for unistd.h - found
-- Looking for vfork.h
-- Looking for vfork.h - not found
-- Looking for zlib.h
-- Looking for zlib.h - found
-- Looking for 4 include files stdlib.h, ..., float.h
-- Looking for 4 include files stdlib.h, ..., float.h - found
-- Performing Test LWS_HAVE_VISIBILITY
-- Performing Test LWS_HAVE_VISIBILITY - Success
-- Found ZLIB: /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib/libz.so (found version "1.2.8") 
zlib include dirs: /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include
zlib libraries: /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib/libz.so
Compiling with SSL support
wolfSSL include dir: /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include
wolfSSL libraries: /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib/libcyassl.so
-- Looking for SSL_CTX_set1_param
-- Looking for SSL_CTX_set1_param - not found
-- Looking for X509_VERIFY_PARAM_set1_host
-- Looking for X509_VERIFY_PARAM_set1_host - not found
Generating API documentation
-- Looking for RPMTools... - rpmbuild NOT FOUND
---------------------------------------------------------------------
  Settings:  (For more help do cmake -LH <srcpath>)
---------------------------------------------------------------------
 LWS_WITH_STATIC = ON
 LWS_WITH_SHARED = ON
 LWS_WITH_SSL = ON (SSL Support)
 LWS_SSL_CLIENT_USE_OS_CA_CERTS = 1
 LWS_USE_WOLFSSL = ON (wolfSSL/CyaSSL replacement for OpenSSL)
   LWS_WOLFSSL_LIBRARIES = /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib/libcyassl.so
   LWS_WOLFSSL_INCLUDE_DIRS = /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include
 LWS_USE_POLARSSL = OFF (PolarSSL replacement for OpenSSL)
 LWS_USE_MBEDTLS = OFF (mbedtls (nee Polarssl) replacement for OpenSSL)
 LWS_WITHOUT_BUILTIN_SHA1 = OFF
 LWS_WITHOUT_BUILTIN_GETIFADDRS = OFF
 LWS_WITHOUT_CLIENT = OFF
 LWS_WITHOUT_SERVER = OFF
 LWS_LINK_TESTAPPS_DYNAMIC = OFF
 LWS_WITHOUT_TESTAPPS = ON
 LWS_WITHOUT_TEST_SERVER = OFF
 LWS_WITHOUT_TEST_SERVER_EXTPOLL = OFF
 LWS_WITHOUT_TEST_PING = OFF
 LWS_WITHOUT_TEST_ECHO = OFF
 LWS_WITHOUT_TEST_CLIENT = OFF
 LWS_WITHOUT_TEST_FRAGGLE = OFF
 LWS_WITHOUT_EXTENSIONS = OFF
 LWS_WITH_LATENCY = OFF
 LWS_WITHOUT_DAEMONIZE = ON
 LWS_USE_LIBEV = 
 LWS_USE_LIBUV = 
 LWS_IPV6 = ON
 LWS_UNIX_SOCK = OFF
 LWS_WITH_HTTP2 = OFF
 LWS_MBED3 = OFF
 LWS_SSL_SERVER_WITH_ECDH_CERT = OFF
 LWS_MAX_SMP = 32
 LWS_WITH_CGI = OFF
 LWS_HAVE_OPENSSL_ECDH_H = 
 LWS_HAVE_SSL_CTX_set1_param = 
 LWS_WITH_HTTP_PROXY = OFF
 LIBHUBBUB_LIBRARIES = 
 PLUGINS = 
 LWS_WITH_ACCESS_LOG = OFF
 LWS_WITH_SERVER_STATUS = OFF
---------------------------------------------------------------------
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_ASM_COMPILER
    CMAKE_ASM_COMPILER_ARG1
    CMAKE_CXX_COMPILER
    CMAKE_CXX_COMPILER_ARG1
    CMAKE_CXX_FLAGS_RELEASE
    CMAKE_EXE_LINKER_FLAGS
    CMAKE_MODULE_LINKER_FLAGS
    DL_LIBRARY


-- Build files have been written to: /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3
rm -f /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.configured_*
touch /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.configured_yyyyyyyy
CFLAGS="-Os -pipe -march=pentium4 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -iremap /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3:libwebsockets-2.0.3 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -I/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include -I/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/include -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/include -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/include/fortify -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/include " CXXFLAGS="-Os -pipe -march=pentium4 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -iremap /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3:libwebsockets-2.0.3 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -I/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include -I/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/include -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/include -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/include/fortify -I/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/include " LDFLAGS="-L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/usr/lib -L/home/main/x86/staging_dir/toolchain-i386_pentium4_gcc-5.3.0_musl-1.1.15/lib -znow -zrelro " make -j1 -C /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/. AR="i486-openwrt-linux-musl-gcc-ar" AS="i486-openwrt-linux-musl-gcc -c -Os -pipe -march=pentium4 -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -iremap /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3:libwebsockets-2.0.3 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" LD=i486-openwrt-linux-musl-ld NM="i486-openwrt-linux-musl-gcc-nm" CC="i486-openwrt-linux-musl-gcc" GCC="i486-openwrt-linux-musl-gcc" CXX="i486-openwrt-linux-musl-g++" RANLIB="i486-openwrt-linux-musl-gcc-ranlib" STRIP=i486-openwrt-linux-musl-strip OBJCOPY=i486-openwrt-linux-musl-objcopy OBJDUMP=i486-openwrt-linux-musl-objdump SIZE=i486-openwrt-linux-musl-size CROSS="i486-openwrt-linux-musl-" ARCH="i386" ;
make[4]: Entering directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
make[5]: Entering directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
make[6]: Entering directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
Scanning dependencies of target websockets
make[6]: Leaving directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
make[6]: Entering directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
[  2%] Building C object CMakeFiles/websockets.dir/lib/base64-decode.c.o
[  4%] Building C object CMakeFiles/websockets.dir/lib/handshake.c.o
[  6%] Building C object CMakeFiles/websockets.dir/lib/libwebsockets.c.o
[  8%] Building C object CMakeFiles/websockets.dir/lib/service.c.o
[ 10%] Building C object CMakeFiles/websockets.dir/lib/pollfd.c.o
[ 13%] Building C object CMakeFiles/websockets.dir/lib/output.c.o
[ 15%] Building C object CMakeFiles/websockets.dir/lib/parsers.c.o
[ 17%] Building C object CMakeFiles/websockets.dir/lib/context.c.o
[ 19%] Building C object CMakeFiles/websockets.dir/lib/alloc.c.o
[ 21%] Building C object CMakeFiles/websockets.dir/lib/header.c.o
[ 23%] Building C object CMakeFiles/websockets.dir/lib/client.c.o
[ 26%] Building C object CMakeFiles/websockets.dir/lib/client-handshake.c.o
[ 28%] Building C object CMakeFiles/websockets.dir/lib/client-parser.c.o
[ 30%] Building C object CMakeFiles/websockets.dir/lib/ssl.c.o
[ 32%] Building C object CMakeFiles/websockets.dir/lib/ssl-server.c.o
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c: In function 'lws_ssl_server_name_cb':
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:160:10: error: 'SSL_TLSEXT_ERR_NOACK' undeclared (first use in this function)
   return SSL_TLSEXT_ERR_NOACK;
          ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:160:10: note: each undeclared identifier is reported only once for each function it appears in
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:163:6: error: implicit declaration of function 'SSL_get_SSL_CTX' [-Werror=implicit-function-declaration]
      SSL_get_SSL_CTX(ssl),
      ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:163:6: error: passing argument 1 of 'wolfSSL_CTX_get_ex_data' makes pointer from integer without a cast [-Werror=int-conversion]
In file included from /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include/cyassl/ssl.h:33:0,
                 from /home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include/cyassl/openssl/ssl.h:30,
                 from /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/private-libwebsockets.h:192,
                 from /home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:22:
/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/usr/include/wolfssl/ssl.h:871:19: note: expected 'const WOLFSSL_CTX * {aka const struct WOLFSSL_CTX *}' but argument is of type 'int'
 WOLFSSL_API void* wolfSSL_CTX_get_ex_data(const WOLFSSL_CTX*, int);
                   ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:173:19: error: comparison between pointer and integer [-Werror]
   if (vh->ssl_ctx == SSL_get_SSL_CTX(ssl))
                   ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:181:15: error: implicit declaration of function 'SSL_get_servername' [-Werror=implicit-function-declaration]
  servername = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
               ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:181:39: error: 'TLSEXT_NAMETYPE_host_name' undeclared (first use in this function)
  servername = SSL_get_servername(ssl, TLSEXT_NAMETYPE_host_name);
                                       ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:188:4: error: implicit declaration of function 'SSL_set_SSL_CTX' [-Werror=implicit-function-declaration]
    SSL_set_SSL_CTX(ssl, vhost->ssl_ctx);
    ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:189:11: error: 'SSL_TLSEXT_ERR_OK' undeclared (first use in this function)
    return SSL_TLSEXT_ERR_OK;
           ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c: In function 'lws_context_init_server_ssl':
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:365:2: error: implicit declaration of function 'SSL_CTX_set_tlsext_servername_callback' [-Werror=implicit-function-declaration]
  SSL_CTX_set_tlsext_servername_callback(vhost->ssl_ctx,
  ^
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c: In function 'lws_ssl_server_name_cb':
/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/lib/ssl-server.c:195:1: error: control reaches end of non-void function [-Werror=return-type]
 }
 ^
cc1: all warnings being treated as errors
CMakeFiles/websockets.dir/build.make:398: recipe for target 'CMakeFiles/websockets.dir/lib/ssl-server.c.o' failed
make[6]: *** [CMakeFiles/websockets.dir/lib/ssl-server.c.o] Error 1
make[6]: Leaving directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/websockets.dir/all' failed
make[5]: *** [CMakeFiles/websockets.dir/all] Error 2
make[5]: Leaving directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
Makefile:149: recipe for target 'all' failed
make[4]: *** [all] Error 2
make[4]: Leaving directory '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3'
Makefile:109: recipe for target '/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.built' failed
make[3]: *** [/home/main/x86/build_dir/target-i386_pentium4_musl-1.1.15/libwebsockets-cyassl/libwebsockets-2.0.3/.built] Error 2
make[3]: Leaving directory '/home/main/x86/feeds/packages/libs/libwebsockets'
package/Makefile:196: recipe for target 'package/feeds/packages/libwebsockets/compile' failed
make[2]: *** [package/feeds/packages/libwebsockets/compile] Error 2
make[2]: Leaving directory '/home/main/x86'
package/Makefile:193: recipe for target '/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/stamp/.package_compile' failed
make[1]: *** [/home/main/x86/staging_dir/target-i386_pentium4_musl-1.1.15/stamp/.package_compile] Error 2
make[1]: Leaving directory '/home/main/x86'

**Comment by lws-team** [comment]
Can you open a new issue for this?

Cyassl official repo at https://github.com/cyassl/cyassl says:

```
"Please use wolfSSL now instead: https://github.com/wolfSSL/wolfssl . CyaSSL is a small, fast, portable implementation of TLS/SSL for embedded devices to the cloud. https://www.wolfssl.com"
```

Let me know on the new issue where exactly your ssl lib came from and exactly what version it is.

**Comment by lws-team** [comment]
I have added some support around this to master a couple of weeks ago along with docs to reproduce.
